### PR TITLE
ci-operator: allow marking post steps as best-effort

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @alvaroaleman 
Implements https://issues.redhat.com/browse/DPTP-1799

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Made this opt-in as I imagine existing post steps like the installer will remain relevant

File: pkg/steps/multi_stage.go [hunk 1]
```diff
@@ -352,7 +355,15 @@ func (s *multiStageTestStep) runSteps(
 const multiStageTestStepContainerName = "test"
 
 func (s *multiStageTestStep) generatePods(steps []api.LiteralTestStep, env []coreapi.EnvVar,
-	hasPrevErrs bool) ([]coreapi.Pod, error) {
+	hasPrevErrs bool) ([]coreapi.Pod, func(string) bool, error) {
+	bestEffort := sets.NewString()
+	isBestEffort := func(podName string) bool {
+		if s.allowBestEffortPostSteps == nil || (s.allowBestEffortPostSteps != nil && !*s.allowBestEffortPostSteps) {
```
Comment (bbguimaraes) [inline pkg/steps/multi_stage.go:361]: Why not just return an empty set from the outer function when this condition is true?

Comment (bbguimaraes) [reply pkg/steps/multi_stage.go:361]:
```suggestion
		if s.allowBestEffortPostSteps == nil || !*s.allowBestEffortPostSteps {
```

File: pkg/api/types.go [hunk 2]
```diff
@@ -735,6 +740,10 @@ type MultiStageTestConfiguration struct {
 	// all previous `pre` and `test` steps were successful. The given step must explicitly
 	// ask for being skipped by setting the OptionalOnSuccess flag to true.
 	AllowSkipOnSuccess *bool `json:"allow_skip_on_success,omitempty"`
+	// AllowBestEffortPostSteps defines if any `post` steps can be ignored when
+	// they fail. The given step must explicitly ask for being ignored by setting
+	// the OptionalOnSuccess flag to true.
+	AllowBestEffortPostSteps *bool `json:"allow_best_effort_post_steps,omitempty"`
```
Comment (bbguimaraes) [inline pkg/api/types.go:746]: Is this flag necessary given that the default for `BestEffort` is `false`?

**Review by bbguimaraes — COMMENTED** [review]
For the record, while I am not opposed to this change, it was part of the original design that the steps would be responsible for managing their exit status.

File: pkg/api/types.go (hunk 2 above)
Comment (stevekuznetsov) [reply pkg/api/types.go:746]: Without it, a step becomes best-effort on all jobs when it's configuration is changed. Is that what we want?

**Comment by stevekuznetsov** [comment]
@bbguimaraes the complication is that we now have per-step timeouts and having a process exit 0 when terminated - especially when the process is in a script, might be hard. This seems simpler.

File: pkg/api/types.go (hunk 2 above)
Comment (bbguimaraes) [reply pkg/api/types.go:746]: Ah, make sense, that's what was trying to understand.

**Pull request merged**